### PR TITLE
 Fixes unnecessary creation of default SG and trying to delete non-provisioned SG by k8s system when annotation [service.beta.kubernetes.io/aws-load-balancer-security-groups] is present

### PR DESCRIPTION
**Pull request by bhks**
**What type of PR is this?**

 /kind bug

**What this PR does / why we need it**:

This PR fixes the Issue : 
- `AWS Cloud Provider attempts to delete LoadBalancer security group it didn’t provision`
- `AWS Cloud Provider creates default LoadBalancer security group even if annotation [service.beta.kubernetes.io/aws-load-balancer-security-groups] is present`

**Which issue(s) this PR fixes**:

Fixes #https://github.com/kubernetes/kubernetes/issues/84256

**Special notes for your reviewer**:

My first PR in Kubernetes

**Does this PR introduce a user-facing change?**:

```release-note

Fixed the following 
-  AWS Cloud Provider attempts to delete LoadBalancer security group it didn’t provision
-  AWS Cloud Provider creates default LoadBalancer security group even if annotation [service.beta.kubernetes.io/aws-load-balancer-security-groups] is present

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

NONE

```

### CODE REVIEW

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go [hunk 1]
```diff
@@ -4301,6 +4301,10 @@ func (c *Cloud) EnsureLoadBalancerDeleted(ctx context.Context, clusterName strin
 		// Delete the security group(s) for the load balancer
 		// Note that this is annoying: the load balancer disappears from the API immediately, but it is still
 		// deleting in the background.  We get a DependencyViolation until the load balancer has deleted itself
+		if securitygroups, ok := service.Annotations[ServiceAnnotationLoadBalancerSecurityGroups]; !ok {
```
Comment (micahhausler) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: tiny nit: can you put a newline above your new code? Just to clarify that the existing comment is not about your specific if block

Comment (bhagwat070919) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: Yeah , make sense Thanks.

I am putting this if block above the comments for the block to separate the block comment for deletion and why if block.

**Review by micahhausler — APPROVED** [review]
/approve
/lgtm
/kind bug

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go (hunk 1 above)
Comment (M00nF1sh) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: 1. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/aws/aws.go#L3484 you will leak security groups..
2. comment "This if code protects"  is not english :D

Comment (bhagwat070919) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: @M00nF1sh Can you explain here how would we leak security group ?

From my understanding we are protecting to call a delete operation on SG which has not been provisioned by k8s.

Comment (bhagwat070919) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: Thanks for pointing out !!!!!!!!

Looking at the Create Code ,

We are creating a security group even if we have the annotation `service.beta.kubernetes.io/aws-load-balancer-security-groups` present and have no value.

I am surprised that we are validating the value but insteading of erroring out  creating a default security group which dissolves the purpose https://kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#load-balancers

**Comment by leakingtapan** [comment]
How is this related to #77557 which is related to "AWS not updating security group for loadBalancerSourceRanges"

And did you address @M00nF1sh 's comment?

**Comment by bhagwat070919** [comment]
@leakingtapan ,

I have updated the Issue Number.

I need to add more code in create step to not create SG when this annotation [service.beta.kubernetes.io/aws-load-balancer-security-groups] is present .

Regards
Bhagwat

**Comment by bhagwat070919** [comment]
## Below is my proposal on this issue which does not solve the problem 100% but tries to fix major problems.
## Problems:
- Even if we have the annotation  `service.beta.kubernetes.io/aws-load-balancer-security-groups` we are creating a new security group.
- We are trying to delete the Security groups provided with annotation `service.beta.kubernetes.io/aws-load-balancer-security-groups` which has not been created by k8s system.

## Solution :

- Stop creating new Security group for load balancer if customers are using annotation `service.beta.kubernetes.io/aws-load-balancer-security-groups`.
- Allow deletion of Security group which has been created previously. May be we can stop deleting from next k8s version.
- Fileter out annotated SG from selection list to not call deletion on the SG not created by k8s system.


This solution does not solve the edge triggered problem but minimizes the impact.

- Customers will not hit the SG limit and API throttling.
- New clusters created would not see the deletion failure if its not edge triggered.


if we are OK with this proposal I can go ahead and make the change and update the PR.

@wongma7 @leakingtapan @micahhausler @M00nF1sh and @jsafrane 

Regards
Bhagwat

**Comment by M00nF1sh** [comment]
> Filter out annotated SG from selection list to not call deletion on the SG not created by k8s system.
This basically means "every SG not specified in the annotation is provisioned by K8s"
Suppose an user create an Service with annotation `SG1`, and an operator manually added `SG2` to that ELB, that SG2 will be deleted when deleting Service.

Instead, I think we should use a way to identify the securityGroups provisioned by K8s directly(e.g. through name or tag).

**Comment by bhagwat070919** [comment]
@M00nF1sh ,

What do you mean by `and an operator manually added SG2 to that ELB` ?
without any annotation ? 

I liked the idea 
`Instead,I think we should use a way to identify the securityGroups provisioned by K8s directly(e.g. through name or tag)` 
but in this case as well there would be a problem where if user adds the same name tag then we will be trying to delete.

but once we make the upstream doc with clear understanding about this tag then it will be much  better.

**Comment by wongma7** [comment]
> Instead,I think we should use a way to identify the securityGroups provisioned by K8s directly(e.g. through name or tag)

This was already added by https://github.com/kubernetes/kubernetes/pull/74311 .

The absence of a cluster tag like `kubernetes.io/cluster/<ClusterID>` on the SG is supposed to mean the SG is ignored and not deleted.

So I am curious how come it doesn't seem to work for the SGs specified in service.beta.kubernetes.io/aws-load-balancer-security-groups.

**Comment by bhagwat070919** [comment]
Added code to perform following :

- Stop creating new Security group for load balancer if customers are using annotation service.beta.kubernetes.io/aws-load-balancer-security-groups.
- Allow deletion of Security group which has been created previously. May be we can stop deleting from next k8s version.
- Fileter out annotated SG from selection list to not call deletion on the SG not created by k8s system.

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go [hunk 2]
```diff
@@ -3520,25 +3538,14 @@ func (c *Cloud) buildELBSecurityGroupList(serviceName types.NamespacedName, load
...
 	if len(sgList) == 0 {
 		sgList = append(sgList, securityGroupID)
 	}
 
-	for _, extraSG := range strings.Split(annotations[ServiceAnnotationLoadBalancerExtraSecurityGroups], ",") {
-		extraSG = strings.TrimSpace(extraSG)
-		if len(extraSG) > 0 {
-			sgList = append(sgList, extraSG)
-		}
+	extraSGList := getSGListFromAnnotation(annotations[ServiceAnnotationLoadBalancerExtraSecurityGroups])
```
Comment (M00nF1sh) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: nit: can be replaced by `sgList = append(sgList, extraSGList...)`

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go [hunk 3]
```diff
@@ -3507,9 +3520,14 @@ func (c *Cloud) buildELBSecurityGroupList(serviceName types.NamespacedName, load
 	var err error
 	var securityGroupID string
 
+	sgList := getSGListFromAnnotation(annotations[ServiceAnnotationLoadBalancerSecurityGroups])
+
+	// The below code changes makes sure that when we have Security Groups  specified with the ServiceAnnotationLoadBalancerSecurityGroups
+	// annotation we don't create a new default Security Groups
+
```
Comment (M00nF1sh) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: nit, while whole part can be reorganized to like to make the logic more clear. But current one is ok too.
```
if len(sgList) == 0 {
    var securityGroupID string
    if c.cfg.Global.ElbSecurityGroup != "" {
       ...
    } else {
        ...
    }
    sgList = append(sgList, securityGroupID)
}
```

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go [hunk 4]
```diff
@@ -4347,6 +4354,11 @@ func (c *Cloud) EnsureLoadBalancerDeleted(ctx context.Context, clusterName strin
 
 		// Collect the security groups to delete
 		securityGroupIDs := map[string]struct{}{}
+		annotatedSgSet := map[string]bool{}
+		annotatedSgsList := getSGListFromAnnotation(service.Annotations[ServiceAnnotationLoadBalancerSecurityGroups])
```
Comment (M00nF1sh) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: how about ServiceAnnotationLoadBalancerExtraSecurityGroups :D

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go (hunk 3 above)
Comment (bhagwat070919) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: Good point , Thanks for the tip.

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go (hunk 2 above)
Comment (bhagwat070919) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: Yeah , I was looking at the Golang doc I couldn't figure out if we have something to just append without own loop.

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go (hunk 4 above)
Comment (bhagwat070919) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: Yes Sure can be and should be added.

**Review by M00nF1sh — APPROVED** [review]

**Review by micahhausler — CHANGES_REQUESTED** [review]
/lgtm
just needs `go fmt`

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go [hunk 5]
```diff
@@ -3497,6 +3497,19 @@ func getPortSets(annotation string) (ports *portSets) {
 	return
 }
 
+// This function is useful in extracting the security group list from annotation
+func getSGListFromAnnotation( annotatedSG string)([]string){
```
[the remark this answers is not in the document]
Comment (bhagwat070919) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: Pushed after running go fmt on 

```
go fmt staging/src/k8s.io/legacy-cloud-providers/aws/aws.go
```

**Review by micahhausler — APPROVED** [review]
/lgtm
/approve

**Pull request merged**